internal/orchestrator/updates.go

```diff
@@ -279,16 +279,159 @@ func parseEventUpdateFromKV(ctx context.Context, kve nats.KeyValueEntry,
+		Target:    evt.ServerID,
+		Kind:      evt.Kind,
+		State:     evt.State,
+		Status:    evt.Status,
+		CreatedAt: evt.CreatedAt,
+		UpdatedAt: evt.UpdatedAt,
+	}
+}
+
+func filterIncompleteRecords(records []*store.ConditionRecord) (map[rctypes.Kind]bool, []*store.ConditionRecord) {
+	foundKinds := map[rctypes.Kind]bool{}
+	foundRecords := []*store.ConditionRecord{}
+
+	for _, cr := range records {
+		if len(cr.Conditions) == 0 {
+			continue
+		}
+
+		if !rctypes.StateIsComplete(cr.State) {
+			foundKinds[cr.Conditions[0].Kind] = true
+			foundRecords = append(foundRecords, cr)
+		}
+	}
+
+	return foundKinds, foundRecords
+}
+
+// Method returns a slice of Conditions along with a slice of ConditionUpdateEvents which
+// are to be applied to the active-condition KV.
+func filterToReconcile(records []*store.ConditionRecord, updateEvts map[string]*v1types.ConditionUpdateEvent) ([]*rctypes.Condition, []*v1types.ConditionUpdateEvent) {
+	currentCRs := map[string]struct{}{}
+
+	// missing conditions in active-conditions KV to be created
+	creates := []*rctypes.Condition{}
+
+	// in-complete status KV updates to be applied
+	updates := []*v1types.ConditionUpdateEvent{}
+
+	stale := func(createdAt, updatedAt time.Time) bool {
+		return time.Since(createdAt) >= rctypes.StaleThreshold &&
+			time.Since(updatedAt) >= rcontroller.StatusStaleThreshold
+	}
+
+	// updates
+	// - The status KV entry does not exist for the condition listed in active-condition condition
+	// - The first incomplete Condition has exceeded the stale threshold
+	for _, cr := range records {
+		firstCond := cr.Conditions[0]
+		currentCRs[firstCond.Target.String()] = struct{}{}
+
+		_, exists := updateEvts[firstCond.Target.String()]
+		if !exists && stale(firstCond.CreatedAt, firstCond.UpdatedAt) {
+			updates = append(updates, failedUpdateEventFromCondition(firstCond))
+		}
+	}
+
+	//  creates
+	// - The non-final Condition Status entry has no corresponding Active Condition Record.
+	for serverID, updateEvt := range updateEvts {
+		_, exists := currentCRs[serverID]
+		if !exists {
+			creates = append(creates, conditionFromUpdateEvent(updateEvt))
+		}
+	}
+
+	return creates, updates
+}
+
+func (o *Orchestrator) activeConditionsToReconcile(ctx context.Context) ([]*rctypes.Condition, []*v1types.ConditionUpdateEvent) {
+	// list condition records from the active-conditions KV
+	records, err := o.repository.List(ctx)
+	if err != nil {
+		o.logger.WithError(err).Error("condition record lookup error")
+	}
+
+	// returned in-complete status KV updates to reconcile
+	updates := []*v1types.ConditionUpdateEvent{}
+
+	// returned missing conditions from active-conditions KV to reconcile
+	creates := []*rctypes.Condition{}
+
+	// filter condition Kinds and records to be reconciled
+	filteredKinds, filteredRecords := filterIncompleteRecords(records)
+
+	// List Conditions in the Status KV and prepare an update payload
+	// for those to be reconciled.
+	for kind := range filteredKinds {
```

Comment:
We don't care what kind of condition it is that needs reconciliation. The criteria are based solely on time and controller liveness. I don't think _de facto_ batching does anything aside increase complexity.

Reply:
fixed in [9616999](https://github.com/metal-toolbox/conditionorc/pull/203/commits/961699928799ae5d6a1d43de442ed848b008fcc3)